File: lib_uikit/src/main/java/com/wujia/lib/widget/StencilFrameLayout.java
package com.wujia.lib.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.widget.FrameLayout;

public class StencilFrameLayout extends FrameLayout {
    Paint mPaint;
    Rect rect;
    Rect innerRect;
    void init(){
        mPaint=new Paint(Paint.ANTI_ALIAS_FLAG);
    }
    public StencilFrameLayout(@NonNull Context context) {
        super(context);
        init();
    }

    public StencilFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public StencilFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public StencilFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }
    private static final int DEFAULT_CHILD_GRAVITY = Gravity.TOP | Gravity.START;
    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        rect=new Rect(left, top, right, bottom);
//        layoutChildren(left, top, right, bottom, false /* no force left gravity */);
    }
    void layoutChildren(int left, int top, int right, int bottom, boolean forceLeftGravity) {
        final int count = getChildCount();
        if(count>1){
            throw new RuntimeException("can not be more than one child");
        }

        final int parentLeft = getPaddingLeft();
        final int parentRight = right - left - getPaddingRight();

        final int parentTop = getPaddingTop();
        final int parentBottom = bottom - top - getPaddingBottom();

        for (int i = 0; i < count; i++) {
            final View child = getChildAt(i);
            if (child.getVisibility() != GONE) {
                final LayoutParams lp = (LayoutParams) child.getLayoutParams();

                final int width = child.getMeasuredWidth();
                final int height = child.getMeasuredHeight();

                int childLeft;
                int childTop;

                int gravity = lp.gravity;
                if (gravity == -1) {
                    gravity = DEFAULT_CHILD_GRAVITY;
                }

                final int layoutDirection = getLayoutDirection();
                final int absoluteGravity = Gravity.getAbsoluteGravity(gravity, layoutDirection);
                final int verticalGravity = gravity & Gravity.VERTICAL_GRAVITY_MASK;

                switch (absoluteGravity & Gravity.HORIZONTAL_GRAVITY_MASK) {
                    case Gravity.CENTER_HORIZONTAL:
                        childLeft = parentLeft + (parentRight - parentLeft - width) / 2 +
                                lp.leftMargin - lp.rightMargin;
                        break;
                    case Gravity.RIGHT:
                        if (!forceLeftGravity) {
                            childLeft = parentRight - width - lp.rightMargin;
                            break;
                        }
                    case Gravity.LEFT:
                    default:
                        childLeft = parentLeft + lp.leftMargin;
                }

                switch (verticalGravity) {
                    case Gravity.TOP:
                        childTop = parentTop + lp.topMargin;
                        break;
                    case Gravity.CENTER_VERTICAL:
                        childTop = parentTop + (parentBottom - parentTop - height) / 2 +
                                lp.topMargin - lp.bottomMargin;
                        break;
                    case Gravity.BOTTOM:
                        childTop = parentBottom - height - lp.bottomMargin;
                        break;
                    default:
                        childTop = parentTop + lp.topMargin;
                }

                child.layout(childLeft, childTop, childLeft + width, childTop + height);
                innerRect=new Rect(childLeft, childTop, childLeft + width, childTop + height);
            }
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
//        mPaint.setAlpha(0);
//        mPaint.setColor(Color.BLACK);
//
////        Rect innerRect=new Rect(child.getLeft(),rect.top+100,rect.right-100,rect.bottom-100);
//        canvas.drawRect(innerRect, mPaint);
        mPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_OUT));

        mPaint.setColor(Color.BLACK);
        mPaint.setAlpha(160);
        canvas.drawRect(rect, mPaint);
        super.dispatchDraw(canvas);//放在super前是后景，相反是前景，前景会覆盖子布局
    }
}

File: lib_uikit/src/main/java/com/wujia/lib/widget/StencilFrameLayout.kt
package com.wujia.lib.widget

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout

class StencilFrameLayout : FrameLayout {
    internal var mPaint: Paint?=null
    private var rect: Rect?=null
    internal var innerRect: Rect?=null
    internal fun init() {
        mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init()
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        rect = Rect(left, top, right, bottom)
        //        layoutChildren(left, top, right, bottom, false /* no force left gravity */);
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
    }

    override fun dispatchDraw(canvas: Canvas) {
        //        mPaint.setAlpha(0);
        //        mPaint.setColor(Color.BLACK);
        //
        ////        Rect innerRect=new Rect(child.getLeft(),rect.top+100,rect.right-100,rect.bottom-100);
        //        canvas.drawRect(innerRect, mPaint);
        mPaint?.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OUT)

        mPaint?.color = Color.BLACK
        mPaint?.alpha = 160
        canvas.drawRect(rect!!, mPaint!!)
        super.dispatchDraw(canvas)//放在super前是后景，相反是前景，前景会覆盖子布局
    }

    companion object {
        private const val DEFAULT_CHILD_GRAVITY = Gravity.TOP or Gravity.START
    }
}
